### PR TITLE
Fix collection mapping to allow subcollection mapping over multi-data inputs.

### LINKED ISSUE #6236: type="data" multiple="true" doesn't allow subcollection mapping over lists
If you map a "list:list" over a ``type="data_collection" collection_type="list"`` tool parameter - it does the correct thing I think and reduces the inner lists while mapping over the outer list (to get collection outputs with identifiers corresponding to the outer list identifiers). It doesn't seem like the multiple data parameters behaves the same way but they really need to.

WIP: https://github.com/jmchilton/galaxy/tree/subcollection_mapping_with_multi_data_inputs

### PR DESCRIPTION
**Pull request by jmchilton**
Calling it a fix with the hope of back porting to release_18.05 at some point - it is jarring to users that works for collection inputs but not multiple data inputs, especially because the IUC discourages collection inputs when multiple data inputs would work.

Fixes #6236 and #5875.

### CODE REVIEW

File: lib/galaxy/tools/parameters/basic.py [hunk 1]
```diff
@@ -1853,12 +1864,17 @@ def append(list, hda, name, src, keep=False):
         # add dataset collections
         dataset_collection_matcher = dataset_matcher_factory.dataset_collection_matcher(dataset_matcher)
         for hdca in history.active_visible_dataset_collections:
-            match = dataset_collection_matcher.hdca_match(hdca, reduction=multiple)
+            match = dataset_collection_matcher.hdca_match(hdca)
             if match:
+                subcollection_type = None
+                if multiple:
+                    subcollection_type = self._history_query(trans).can_map_over(hdca).collection_type
```
Comment (mvdbeek) [inline lib/galaxy/tools/parameters/basic.py]: Something isn't quite right here, I see this traceback when building the tool form:

```
galaxy.web.framework.decorators ERROR 2018-06-01 10:45:31,618 Uncaught exception in exposed API method:
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/web/framework/decorators.py", line 281, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/webapps/galaxy/api/tools.py", line 113, in build
    return tool.to_json(trans, kwd.get('inputs', kwd))
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/__init__.py", line 1883, in to_json
    self.populate_model(request_context, self.inputs, state_inputs, tool_model['inputs'])
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/__init__.py", line 1954, in populate_model
    tool_dict = input.to_dict(request_context)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/tools/parameters/basic.py", line 1871, in to_dict
    subcollection_type = self._history_query(trans).can_map_over(hdca).collection_type
AttributeError: 'bool' object has no attribute 'collection_type'
```

Comment (jmchilton) [reply lib/galaxy/tools/parameters/basic.py]: Yes - the other place this is used in that file it verifies it is truthy before using it - my bad. I'll throw in a continue in that loop if it is false.

Comment (mvdbeek) [reply lib/galaxy/tools/parameters/basic.py]: I see you've rebased the commit, but something must've gone wrong again.
I suppose this should be as in 880c0217ab3d007e34431c6ae54101f89c1aa3b4 ?

**Comment by mvdbeek** [comment]
I'm not sure this does the right thing (all the time). If I use the `multi_data_param` tool and I feed it twice the same nested collection it completely reduces the collections to 2 regular HDAs, while I would have expected to just reduce the deepest list.

So in the screenshot I've used collection 13 twice as input to the `multi_data_param` tool and I got the upper 2 HDAs as a result.

screen shot 2018-06-02 at 10 52 03

But maybe I'm thinking about this the wrong way and multiple="true" should always completely reduce entire collections. What I thought would happen I could always achieve with a collection input and a regular output, right ?

**Comment by jmchilton** [comment]
Clearly my rebase was wrong, sorry about. I'll bring in 880c0217ab3d007e34431c6ae54101f89c1aa3b4 - that looks right.

I tried two inputs at once with the same tool and it works for me I think - any chance you didn't rebuild the client? What you are describing is what happens without the client being rebuilt. If you rebuilt the client I'll do some more testing.

Also - since you are doing deep testing I should mention the form doesn't do any validation. So like it makes sense that picking multiple lists reduces all the lists together at once and that should work or that picking a nested list does this subcollection mapping thing you expect - but if you mix and match the two or specify multiple nested lists ... things are going to break and the sharp change of behavior has no representation in UI between lists and nested lists (this PR mirrors the problem https://github.com/galaxyproject/galaxy/issues/2467 with subcollection mapping over collection paramaeters).

**Comment by mvdbeek** [comment]
Cool! I saw the client being rebuilt, but my chrome is sometimes a bit stubborn with getting the new client if not opening the console and disabling the cache.
So ... fantastic, nested collection work 100% the way I was expecting it would (but I had to include https://github.com/galaxyproject/galaxy/pull/6253/commits/50c98701b7682d8878cbc2fa1acdc3f28abdad0e).

**Review by mvdbeek — APPROVED** [review]
Super awesome, thanks a lot @jmchilton !

**Pull request merged**